src/server/scripts/EasternKingdoms/Stratholme/instance_stratholme.cpp

```diff
@@ -474,91 +573,118 @@ class instance_stratholme : public InstanceMapScript
+                    break;
+                }
+                case EVENT_BARTHILAS_MOVE_END:
+                    if (Creature* barthilas = instance->GetCreature(_barthilasGUID))
                     {
-                        Map::PlayerList const& PlayerList = instance->GetPlayers();
-                        if (!PlayerList.IsEmpty())
-                            for (Map::PlayerList::const_iterator i = PlayerList.begin(); i != PlayerList.end(); ++i)
-                                if (Player* player = i->GetSource())
-                                    if (player->GetDistance2d(4034.97f, -3402.13f) < 50.0f)
-                                    {
-                                        ProcessSlaughterEvent();
-                                        return;
-                                    }
-
-                        events.ScheduleEvent(EVENT_FORCE_SLAUGHTER_EVENT, 3s);
-                        break;
+                        if (barthilas->isDead())
+                            break;
+                        if (barthilas->IsAlive() && (barthilas->GetPosition()) == Position(4068.28f, -3535.67f, 122.771f, 2.50f))
+                        {
+                            barthilas->SetHomePosition(4068.28f, -3535.67f, 122.771f, 2.50f);
+                            SetData(TYPE_BARTHILAS_RUN, DONE);
+                            SaveToDB();
+                            break;
+                        }
                     }
-                case EVENT_SPAWN_BLACK_GUARD:
+                    events.ScheduleEvent(EVENT_BARTHILAS_MOVE_END, 4s);
+                    break;
+                case EVENT_BARTHILAS_TEL_TO:
+                    if (Creature* barthilas = instance->GetCreature(_barthilasGUID))
                     {
-                        for (uint8 i = 0; i < 5; ++i)
-                            if (Creature* guard = instance->SummonCreature(NPC_BLACK_GUARD, BlackGuardPos[i]))
-                            {
-                                guard->SetWalk(true);
-                                guard->GetMotionMaster()->MovePoint(0, BlackGuardPos[i + 5]);
-                                guard->SetHomePosition(BlackGuardPos[i + 5]);
-                                if (i == 0 && guard->AI())
-                                    guard->AI()->Talk(SAY_BLACK_GUARD_INIT);
-                            }
-
-                        if (GameObject* gate = instance->GetGameObject(_zigguratDoorsGUID4))
-                            gate->SetGoState(GO_STATE_ACTIVE);
-                        break;
+                        if (barthilas->IsAlive() && !barthilas->IsInCombat())
+                        {
+                            barthilas->NearTeleportTo(4068.284f, -3535.678f, 122.771f, 2.50f);
+                            barthilas->SetHomePosition(4068.284f, -3535.678f, 122.771f, 2.50f);
+                        }
                     }
+                    break;
+                default:
+                    break;
             }
```

Comment:
This transition looks pretty abrupt in-game -- he disappears immediately at the game, no fade out. I tried to alter some of this code to get a nice fade out transition (using `barthilas->SetVisible(false)`) but didn't have any luck getting it to work. Maybe you have some better ideas?

I do wish you could get a dynamic trigger here, a la `BossAI`:
https://github.com/azerothcore/azerothcore-wotlk/blob/master/src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp#L146-L161

But I don't think that's available here.

Finally, can you please move this position data up top with the other Position definitions?

Reply:
Ah! I see how you're moving him now with a `waypoint_script`, which is causing this little hiccup.

Would you consider taking out the teleport with waypoint script and instead schedule the event `EVENT_BARTHILAS_TEL_TO`? Then add a `SetVisible(false)` before and `SetVisible(true)` after the teleport in the event?

That should reliably get him there and won't create the pop-out effect anymore. Thoughts?

Reply:
Well, turns out this matches exactly how Classic works - he just disappears abruptly. I confirmed via sniff that he's doing a teleport, not a spell. So your implementation is correct!